controllers/openshift_controller.go

```diff
@@ -1210,12 +1183,151 @@ func (eh *McpEventHandler) Generic(event event.GenericEvent, queue workqueue.Rat
 }
 
 
+func (r *KataConfigOpenShiftReconciler) nodeMatchesKataSelector(nodeLabels map[string]string) bool {
+	nodeSelector, err := r.getNodeSelectorAsSelector()
+
+	if err != nil {
+		r.Log.Info("couldn't get kata node selector", "err", err)
+		// If we cannot find out whether a Node matches assuming that it
+		// doesn't seems to be the safer assumption.  This also seems
+		// consistent with error-handling sematics of earlier similar code.
+		return false
+	}
+
+	return nodeSelector.Matches(labels.Set(nodeLabels))
+}
+
+func isWorkerNode(node client.Object) bool {
+	if _, ok := node.GetLabels()["node-role.kubernetes.io/worker"]; ok {
+		return true
+	}
+	return false
+}
+
+func getStringMapDiff(oldMap, newMap map[string]string) (added, modified, removed map[string]string) {
+	added = map[string]string{}
+	modified = map[string]string{}
+	removed = map[string]string{}
+
+	if oldMap == nil {
+		added = newMap
+		return
+	}
+
+	if newMap == nil {
+		removed = oldMap
+		return
+	}
+
+	for newKey, newVal := range newMap {
+		if _, ok := oldMap[newKey]; !ok {
+			added[newKey] = newVal
+		}
+	}
+	for oldKey, oldVal := range oldMap {
+		if _, ok := newMap[oldKey]; !ok {
+			removed[oldKey] = oldVal
+		} else {
+			if oldMap[oldKey] != newMap[oldKey] {
+				modified[oldKey] = newMap[oldKey]
+			}
+		}
+	}
+	return added, modified, removed
+}
+
+
+type NodeEventHandler struct {
+	reconciler *KataConfigOpenShiftReconciler
+}
+
+func (eh *NodeEventHandler) Create(event event.CreateEvent, queue workqueue.RateLimitingInterface) {
+	node := event.Object
+
+	log := eh.reconciler.Log.WithName("NodeCreate").WithValues("node name", node.GetName())
+	log.Info("node created")
+
+	if !isWorkerNode(node) {
+		return
+	}
+
+	if eh.reconciler.kataConfig == nil {
+		return
+	}
+
+	if !eh.reconciler.nodeMatchesKataSelector(node.GetLabels()) {
+		return
+	}
+	log.Info("node matches kata node selector", "node labels", node.GetLabels())
+
+	queue.Add(eh.reconciler.makeReconcileRequest())
+}
+
+func (eh *NodeEventHandler) Update(event event.UpdateEvent, queue workqueue.RateLimitingInterface) {
+	// This function assumes that a node cannot change its role from master to
+	// worker or vice-versa.
+	nodeOld := event.ObjectOld
+	nodeNew := event.ObjectNew
+
+	log := eh.reconciler.Log.WithName("NodeUpdate").WithValues("node name", nodeNew.GetName())
+
+	if !isWorkerNode(nodeNew) {
+		return
+	}
+
+	if eh.reconciler.kataConfig == nil {
+		return
+	}
+
+	foundRelevantChange := false
+
+	// no need to check the second return value of the indexing operation
+	// as "" is not a valid machineconfiguration.openshift.io/state value
+	stateOld := nodeOld.GetAnnotations()["machineconfiguration.openshift.io/state"]
+	stateNew := nodeNew.GetAnnotations()["machineconfiguration.openshift.io/state"]
+	if stateOld != stateNew {
+		foundRelevantChange = true
```

Comment:
At this point, you can just queue the reconcile request and return right?
There is no need to compare the labels?

Reply:
For bare functionality, yes you're right. But we want the complete logs, trust me. ;-D  I found it extremely helpful to know for a fact what the states of individual nodes are when trying to understand or debug something, esp. from an incomplete log file where you cannot check the earlier messages for information that you're missing.